README.md
# clj-quakes

FIXME: description

## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar clj-quakes-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2017 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-11-27
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-11-27
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/clj-quakes/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/clj-quakes/compare/0.1.0...0.1.1


test/clj_quakes/core_test.clj
(ns clj-quakes.core-test
  (:require [clojure.test :refer :all]
            [clj-quakes.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/clj_quakes/core.clj
(ns clj-quakes.core
  (:gen-class)
  (:require [clj-quakes.quakes :as quakes]
            [clj-http.client :as client]
            [clojure.java.io :refer [resource file]] ;; for testing
            [clojure.core.async :as async  :refer [>! <! >!! <!! go chan buffer close! thread alts! alts!! timeout]]
            [clojure.tools.logging :as log]
            [clojurewerkz.quartzite.jobs :as j]
            [clojurewerkz.quartzite.jobs :refer [defjob]]
            [clojurewerkz.quartzite.schedule.cron :refer [schedule cron-schedule]]
            [clojurewerkz.quartzite.scheduler :as qs]
            [clojurewerkz.quartzite.triggers :as t]
            [cheshire.core :refer :all]
            [com.climate.geojson-schema.core :refer [FeatureCollection GeoJSON]]
            [clojurewerkz.machine-head.client :as mh]
            [schema.core :as s]))

(def ^:const owntracks-topic "owntracks/#")
(def ^:const broker-url "tcp://sharkbaitextraordinaire.com:8885") ;; figure out how to use a TLS-secured connection

(def hood
  {:type        "Point"
   :coordinates [-121.695728,
                 45.37476]})
(def helens
  {:type "Point"
   :coordinates [-122.189941,
                  46.197419]})

(def monitored-locations (list hood helens))

(defn parse-owntracks
  "Parse an mqtt payload into a map with keys"
  [^bytes payload]
  (let [update
        (cheshire.core/parse-string (String. payload "UTF-8") true)]))

(defjob FetchJob
  [ctx]
  (comment "Fetch quakes, filter interesting/worrisome ones, and deal with them")
  (let [quakes (clj-quakes.quakes/newer? (:features clj-quakes.quakes/fetch))]
  ; filter out older quakes
  ; filter out smaller quakes
  ; filter out quakes farther than "interesting" km away 
  ; filter out quakes closer than "worrisome" km away
  ; do something for each "interesting" quake
  ; do something else for each "worrisome" quake
)
  ; This expression returns quakes that occurred within the past 6 minutes and within 1000km:
  ;(clj-quakes.quakes/nearness-filter (clj-quakes.quakes/new? (:features (clj-quakes.quakes/fetch))) clj-quakes.quakes/test-point 1000)
  ; we should also filter the quakes using a closer nearness-filter and a magnitude filter?
  ; or do like in the java version and just drop the fetched earthquakes onto a queue and analyze them later
  ; analysis side would pop quakes off queue, determine closest monitored location to quake,
  ; and then determine if it was 'interesting' or 'worrisome'
  ; then drop the interesting/worrisome quakes onto a queue for outbound notifications
  (println "fetch job is running")) ; we should log/warn this instead of println

;; TODO use command line arguments or parse env/yaml to load configuration
(defn -main
  "I don't do a whole lot ... yet."
  [& args]
  (println "Hello, World!")

  ;; using quartzite to run the fetch job evey five minutes
  (let [s (-> (qs/initialize) qs/start)
        job (j/build
              (j/of-type FetchJob)
              (j/with-identity (j/key "jobs.quakes.fetcher")))
        trigger (t/build
                 (t/with-identity (t/key "trigger.fetch.1"))
                 (t/start-now)
                 (t/with-schedule (schedule
                                   (cron-schedule "0 0/5 * * * ?	"))))]
    (qs/schedule s job trigger))

;  (let [id   (mh/generate-id)
;        conn (mh/connect broker-url id)]
;    (mh/subscribe conn {owntracks-topic 0}
;                  (fn [^String topic meta ^bytes payload]
;                    (parse-owntracks payload))))
  )


;; This outputs the distance from the test point (in km) for every earthquake present in the feed
;; needs improvement by also including the location and detail url
;; and next step is to determine if the elements are within a range to be considered interesting
;; (map distance-from-test (map :geometry (:features (fetch))))


src/clj_quakes/fetcher.clj
(ns clj-quakes.fetcher
  (:require [clj-http.client :as client]
            [clojure.java.io :refer [resource file]] ;; for testing
            [clojure.tools.logging :as log]
            [com.climate.geojson-schema.core :refer [FeatureCollection GeoJSON]]
            [schema.core :as s]))

;; TODO extract to tests package
(defn test-feed []
  (s/validate com.climate.geojson-schema.core/FeatureCollection
              (cheshire.core/parse-string
                (slurp "/Users/vigevenoj/code/clojure/clj-quakes/hourly.json") true)))

;; use this as a test point for haversine and such. It's in Portland.
;; TODO extract to tests package
(def test-point
  {:type        "Point"
   :coordinates [-122.58441925048828,
                 45.48420633926945]})

(defn get-coordinates [feature]
  "Get the coordinates of a feature"
  (:coordinates (:geometry feature)))

(defn get-longitude
  "Get the longitude from a GeoJSON Point"
  [point]
  (first (:coordinates point)))

(defn get-latitude
  "Get the latitude from a GeoJSON Point"
  [point]
  (first (rest (:coordinates point))))

(defn hav
  "Calculate a haversine, sin^2(θ/2)"
  [val]
  (Math/pow (Math/sin (/ val 2)) 2))

(defn point-haversine
  "Haversine formula to calculate great-circle distance between two GeoJSON Point objects"
  [p1 p2]
  (let [R               6371 ; radius of arth in km
        delta-latitude  (Math/toRadians (- (get-latitude p1) (get-latitude p2)))
        delta-longitude (Math/toRadians (- (get-longitude p1) (get-longitude p2)))
        latitude1       (Math/toRadians (get-latitude p1))
        latitude2       (Math/toRadians (get-latitude p2))
        a               (+ (hav delta-latitude)
                           (* (Math/cos latitude1) (Math/cos latitude2) (hav delta-longitude)))]
    (* R 2 (Math/atan2 (Math/sqrt a) (Math/sqrt (- 1 a))))))

(defn fetch-quakes []
  "Get the earthquake feed and parse it as a FeatureCollection"
  (let [response
        (cheshire.core/parse-string
         (:body
           (client/get "https://earthquake.usgs.gov/earthquakes/feed/v1.0/summary/all_hour.geojson"))
         true)]
    (s/validate FeatureCollection response)))

(defn distance-from-test
  "Calculate how far a point is from the test point"
  [point]
  (point-haversine test-point point))

(defn newness-filter
  "Filter out quakes older than five minutes"
  [quakes]
  ;; should be true if newer than 1000*60*5
  (filter #(<
            (- (System/currentTimeMillis) (-> % :properties :time))
            (* 24 60 60 1000)) ; 60 minutes in milliseconds
          quakes))

;; this prints out the distance from the test point to each earthquake
;; we should use this for calculating if the earthquake is close enough to care about
(map clj-quakes.fetcher/distance-from-test (->> (:features (clj-quakes.fetcher/test-feed)) (map :geometry)))

;(defn millis-ago [timestamp] (- (System/currentTimeMillis) timestamp))
;(map millis-ago ((:features (clj-quakes.fetcher/test-feed)) (map :properties) (map :time)))
; below: filter returns collection of points where points are less than 1000km from test-point
;(filter #(< (clj-quakes.fetcher/distance-from-test %) 1000) (->> (:features (clj-quakes.fetcher/test-feed)) (map :geometry)))
; below: filter returns collection of quakes (maps) where points are less than 1000km from test-point
;(filter #(< (clj-quakes.fetcher/distance-from-test (:geometry %)) 1000) (:features (clj-quakes.fetcher/test-feed)) )
(defn nearness-filter
  [quakes location distance]
  (filter #(<
            (clj-quakes.fetcher/point-haversine location (:geometry %))
            distance)
          quakes ))

src/clj_quakes/quakes.clj
(ns clj-quakes.quakes
  (:require [clj-http.client :as client]
            [clojure.java.io :refer [resource file]] ;; for testing
            [clojure.tools.logging :as log]
            [com.climate.geojson-schema.core :refer [FeatureCollection GeoJSON]]
            [schema.core :as s]))

;; TODO extract to tests package
(defn test-feed []
  (s/validate com.climate.geojson-schema.core/FeatureCollection
              (cheshire.core/parse-string
                (slurp  "hourly.json") true)))

;; use this as a test point for haversine and such. It's in Portland.
;; TODO extract to tests package
(def test-point
  {:type        "Point"
   :coordinates [-122.58441925048828,
                 45.48420633926945]})

(defn get-coordinates [feature]
  "Get the coordinates of a feature"
  (:coordinates (:geometry feature)))

(defn get-longitude
  "Get the longitude from a GeoJSON Point"
  [point]
  (first (:coordinates point)))

(defn get-latitude
  "Get the latitude from a GeoJSON Point"
  [point]
  (first (rest (:coordinates point))))

(defn hav
  "Calculate a haversine, sin^2(θ/2)"
  [val]
  (Math/pow (Math/sin (/ val 2)) 2))

(defn point-haversine
  "Haversine formula to calculate great-circle distance between two GeoJSON Point objects"
  [p1 p2]
  (let [R               6371 ; radius of arth in km
        delta-latitude  (Math/toRadians (- (get-latitude p1) (get-latitude p2)))
        delta-longitude (Math/toRadians (- (get-longitude p1) (get-longitude p2)))
        latitude1       (Math/toRadians (get-latitude p1))
        latitude2       (Math/toRadians (get-latitude p2))
        a               (+ (hav delta-latitude)
                           (* (Math/cos latitude1) (Math/cos latitude2) (hav delta-longitude)))]
    (* R 2 (Math/atan2 (Math/sqrt a) (Math/sqrt (- 1 a))))))

(defn fetch []
  "Get the earthquake feed and parse it as a FeatureCollection"
  (let [response
        (cheshire.core/parse-string
         (:body
           (client/get "https://earthquake.usgs.gov/earthquakes/feed/v1.0/summary/all_hour.geojson"))
         true)]
    (s/validate FeatureCollection response)))

;; TODO move this into test package?
(defn distance-from-test
  "Calculate how far a point is from the test point"
  [point]
  (point-haversine test-point point))

(defn newer?
  "Filter out quakes older than six minutes"
  [quakes]
  ;; should be true if newer than 1000*60*5
  (filter #(<
            (- (System/currentTimeMillis) (-> % :properties :time))
            (* 6 60 1000)) ; 6 minutes in milliseconds
          quakes))

;; this prints out the distance from the test point to each earthquake
;; we should use this for calculating if the earthquake is close enough to care about
(map clj-quakes.quakes/distance-from-test (->> (:features (clj-quakes.quakes/test-feed)) (map :geometry)))

;(defn millis-ago [timestamp] (- (System/currentTimeMillis) timestamp))
;(map millis-ago ((:features (clj-quakes.quakes/test-feed)) (map :properties) (map :time)))
; below: filter returns collection of points where points are less than 1000km from test-point
;(filter #(< (clj-quakes.quakes/distance-from-test %) 1000) (->> (:features (clj-quakes.quakes/test-feed)) (map :geometry)))
; below: filter returns collection of quakes (maps) where points are less than 1000km from test-point
;(filter #(< (clj-quakes.quakes/distance-from-test (:geometry %)) 1000) (:features (clj-quakes.quakes/test-feed)) )
(defn nearness-filter
  "filter for quakes closer than a specified distance"
  [quakes location distance]
  (filter #(<
            (clj-quakes.quakes/point-haversine location (:geometry %))
            distance)
          quakes ))

(defn magnitude-filter
  "filter for quakes bigger than a given magnitude"
  [quakes] [mag]
  (filter #(>
            (-> % :properties :mag)
            mag)
          quakes))

(defn closest
  "Calculate minimum distance between a quake and the monitored locations"
  [locations quake]
  (let [c (partial quakes/point-haversine (-> quake :geometry))]
    (map c locations)))

(defn quake-distances
  "Calculate distance between locations and quakes"
  [locations quakes]
  (let [d (partial closest locations)]
    (map d quakes)))